hwtracer/src/llvm_blockmap.rs

```diff
@@ -0,0 +1,216 @@
+//! Parser for ykllvm's extended `.llvm_bb_addr_map`.
+
+use byteorder::{NativeEndian, ReadBytesExt};
+use intervaltree::IntervalTree;
+use libc::{dlsym, RTLD_DEFAULT};
+use std::{
+    convert::TryFrom,
+    ffi::CString,
+    io::{prelude::*, Cursor, SeekFrom},
+    slice,
+    sync::LazyLock,
+};
+
+pub static LLVM_BLOCK_MAP: LazyLock<BlockMap> = LazyLock::new(BlockMap::new);
+
+#[derive(Debug)]
+pub enum SuccessorKind {
+    // FIXME notes on divergence.
+    // One successor.
+    Unconditional {
+        target: Option<u64>,
+    },
+    // Choice of two successors.
+    Conditional {
+        taken_target: u64,
+        not_taken_target: Option<u64>,
+    },
+    // A control flow edge known only at runtime, e.g. a return or an indirect branch.
+    Dynamic,
+}
+
+/// Information about an machine-level LLVM call instruction.
+#[derive(Debug, PartialEq, PartialOrd)]
+pub struct CallInfo {
+    /// Offset of the call instruction
+    callsite_off: u64,
+    /// Offset of the target of the call (if known statically).
+    target_off: Option<u64>,
+}
+
+impl CallInfo {
+    pub fn callsite_off(&self) -> u64 {
+        self.callsite_off
+    }
+
+    pub fn target_off(&self) -> Option<u64> {
+        self.target_off
+    }
+}
+
+/// The information for one LLVM `MachineBasicBlock`.
+#[derive(Debug)]
+pub struct BlockMapEntry {
+    /// Function offset.
+    f_off: u64,
+    /// Indices of corresponding BasicBlocks.
+    corr_bbs: Vec<u64>,
+    /// Successor information.
+    succ: SuccessorKind,
+    /// FIXME: doc
```

Comment:
And another.

Reply:
Fixed in 1bb4a8c279fe9d7ad4295c13dac7ef71935fe20c